src/server/protocol.ts

```diff
@@ -395,17 +398,63 @@ namespace ts.server.protocol {
     }
 
     /**
+     * An diagnostic information suggesting refactors at applicable positions without
+     * clients asking.
+     */
+    export interface RefactorDiagnostic {
+        text: string;
+        code: number;
+        start: Location;
+        end: Location;
+    }
+
+    export interface RefactorDiagnosticEventBody {
+        file: string;
+        diagnostics: RefactorDiagnostic[];
+    }
+
+    /**
+     * Returns a list of applicable refactors at a given position. This request does not actually
+     * compute the refactors; instead it goes through faster checks to determine what is possible.
+     */
+    export interface GetRefactorsForRangeRequest extends Request {
+        command: CommandTypes.GetRefactorsForRange;
+        arguments: GetRefactorsForRangeRequestArgs;
+    }
+
+    export interface GetRefactorsForRangeRequestArgs extends TextRangeRequestArgs {
+    }
+
+    export interface GetRefactorsForRangeResponse extends Response {
+        body?: RefactorDiagnostic[];
+    }
+
+    /**
+     * Computes the code actions for a given refactor. This is normally called after the user commited one
+     * of the applicable refactors provided by the "GetApplicableRefactors" API.
+     */
+    export interface GetCodeActionsForRefactorRequest extends Request {
+        command: CommandTypes.GetCodeActionsForRefactor;
+        arguments: GetCodeActionsForRefactorRequestArgs;
+    }
+
+    export interface GetCodeActionsForRefactorRequestArgs extends GetRefactorsForRangeRequestArgs {
+        refactorCode: number;
+    }
+
+    export interface GetCodeActionsForRefactorResponse extends Response {
+        body?: CodeAction[];
+    }
+
+    /**
       * Request for the available codefixes at a specific position.
       */
     export interface CodeFixRequest extends Request {
         command: CommandTypes.GetCodeFixes;
         arguments: CodeFixRequestArgs;
     }
 
-    /**
-      * Instances of this interface specify errorcodes on a specific location in a sourcefile.
-      */
-    export interface CodeFixRequestArgs extends FileRequestArgs {
+    export interface TextRangeRequestArgs extends FileRequestArgs {

```

Comment:
Maybe `FileRangeRequestArgs` would fit better. But using `location` for start and end would be an unnecessary breaking change for editors already consuming this type.